src/wasm/asm-wasm-builder.cc

```diff
@@ -453,26 +574,108 @@ class AsmWasmBuilderImpl : public AstVisitor {
   }                                                                   \
   break
 
+  Expression* GetLeft(BinaryOperation* expr) {
+    if (expr->op() == Token::BIT_XOR) {
+      return expr->left()->AsBinaryOperation()->left();
+    } else {
+      return expr->left();
+    }
+  }
+
   void VisitBinaryOperation(BinaryOperation* expr) {
-    switch (expr->op()) {
-      BINOP_CASE(Token::ADD, Add, NON_SIGNED_BINOP, true);
-      BINOP_CASE(Token::SUB, Sub, NON_SIGNED_BINOP, true);
-      BINOP_CASE(Token::MUL, Mul, NON_SIGNED_BINOP, true);
-      BINOP_CASE(Token::DIV, Div, SIGNED_BINOP, false);
-      BINOP_CASE(Token::BIT_OR, Ior, NON_SIGNED_INT_BINOP, true);
-      BINOP_CASE(Token::BIT_XOR, Xor, NON_SIGNED_INT_BINOP, true);
-      BINOP_CASE(Token::SHL, Shl, NON_SIGNED_INT_BINOP, true);
-      BINOP_CASE(Token::SAR, ShrS, NON_SIGNED_INT_BINOP, true);
-      BINOP_CASE(Token::SHR, ShrU, NON_SIGNED_INT_BINOP, true);
-      case Token::MOD: {
+    ConvertOperation convertOperation = TestBinaryOperation(expr);
+    if (convertOperation == TO_DOUBLE) {
+      int type = TypeIndexOf(expr->left());
+      if (type == 0) {
+        current_function_builder_->AppendCode(kExprF64SConvertI32, false);
+      } else if (type == 1) {
+        current_function_builder_->AppendCode(kExprF64UConvertI32, false);
+      } else if (type == 2) {
+        current_function_builder_->AppendCode(kExprF64ConvertF32, false);
+      } else {
         UNREACHABLE();
       }
-      break;
-      default:
+      RECURSE(Visit(expr->left()));
+    } else if (convertOperation == TO_INT) {
+      int type = TypeIndexOf(GetLeft(expr));
+      if (type == 2) {
+        current_function_builder_->AppendCode(kExprI32SConvertF32, false);
+      } else if (type == 3) {
+        current_function_builder_->AppendCode(kExprI32SConvertF64, false);
+      } else {
         UNREACHABLE();
+      }
+      RECURSE(Visit(GetLeft(expr)));
+    } else if (convertOperation == AS_IS) {
+      RECURSE(Visit(GetLeft(expr)));
+    } else {
+      bool visit_children = true;
+      switch (expr->op()) {
+        BINOP_CASE(Token::ADD, Add, NON_SIGNED_BINOP, true);
+        BINOP_CASE(Token::SUB, Sub, NON_SIGNED_BINOP, true);
+        BINOP_CASE(Token::MUL, Mul, NON_SIGNED_BINOP, true);
+        BINOP_CASE(Token::DIV, Div, SIGNED_BINOP, false);
+        BINOP_CASE(Token::BIT_OR, Ior, NON_SIGNED_INT_BINOP, true);
+        BINOP_CASE(Token::BIT_XOR, Xor, NON_SIGNED_INT_BINOP, true);
+        BINOP_CASE(Token::SHL, Shl, NON_SIGNED_INT_BINOP, true);
+        BINOP_CASE(Token::SAR, ShrS, NON_SIGNED_INT_BINOP, true);
+        BINOP_CASE(Token::SHR, ShrU, NON_SIGNED_INT_BINOP, true);
+        case Token::MOD: {
+          int type = TypeIndexOf(expr->left(), expr->right(), false);
+          if (type == 0) {
```

Comment:
Can we have an enum here, now that we are testing the result of this TypeIndexOf?